### PR TITLE
Add sm2135 component

### LINKED ISSUE #443: SM2135 support for smart led
**Describe the problem you have/What new integration you would like**
I bought some LSC GU10 Smart LED from Action and would like to use ithem with ESPHome. I've successfully used Tuya-convert but it seems that support for SM2135 is missing from ESPHome

**Please describe your use case for this integration and alternatives you've tried:**
It seems that some smart bulb are using this chip. Would be nice to use ESPHome on it.

**Additional context**
See this issue at Tasmota: https://github.com/arendst/Sonoff-Tasmota/issues/6495
A lot of information can be found there

### PR DESCRIPTION
**Pull request by BoukeHaarsma23**
# What does this implement/fix? 

This implements a SM2135 component. I found the code on @NURDspace repo. However, apparently he did not find the time to create a PR. That's why I'm creating this one. I removed some unused includes and this has been running for almost half a year for me. So I think it would be safe to merge.

I would like to take this opportunity to return something for the community here. So please let me know if something needs to be improved 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/443

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1127
  
# Test Environment

- [ ] ESP32
- [x] ESP8266
- [ ] Windows
- [ ] Mac OS
- [x] Linux

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
output:
  - platform: sm2135
    id: output_red
    channel: 0
  - platform: sm2135
    id: output_green
    channel: 1
    max_power: 0.46
  - platform: sm2135
    id: output_blue
    channel: 2
    max_power: 0.3
  - platform: sm2135
    id: output_white
    channel: 3
    max_power: 0.8
  - platform: sm2135
    id: output_warmwhite
    channel: 4
    max_power: 0.8
```

# Explain your changes

Implements a SM2135 support. Has been working fine for me with the color_interlock feature.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/sm2135/__init__.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome import pins
+from esphome.const import (
+    CONF_CLOCK_PIN,
+    CONF_DATA_PIN,
+    CONF_ID,
+)
+
+AUTO_LOAD = ["output"]
```
Comment (jesserockz) [inline esphome/components/sm2135/__init__.py:10]: Please add yourself as a Codeowner and run `script/build_codeowners.py`

```suggestion
AUTO_LOAD = ["output"]
CODEOWNERS = ["@BoukeHaarsma23"]

```

File: esphome/components/sm2135/sm2135.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  this->data_pin_->digital_write(true);
+  this->clock_pin_->setup();
+  this->clock_pin_->digital_write(true);
+  this->pwm_amounts_.resize(5, 0);
+}
+void SM2135::dump_config() {
+  ESP_LOGCONFIG(TAG, "SM2135:");
+  LOG_PIN("  Data Pin: ", this->data_pin_);
+  LOG_PIN("  Clock Pin: ", this->clock_pin_);
+  ESP_LOGCONFIG(TAG, "  Total number of channels: 5");
```
Comment (jesserockz) [inline esphome/components/sm2135/sm2135.cpp:48]: Static value here?

File: esphome/components/sm2135/sm2135.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      this->write_byte_(buffer[i]);
+    }
+    this->clock_pin_->digital_write(false);
+    this->clock_pin_->digital_write(true);
+    this->data_pin_->digital_write(true);
+  }
+
+  GPIOPin *data_pin_;
+  GPIOPin *clock_pin_;
+  uint8_t num_channels_;
```
Comment (jesserockz) [inline esphome/components/sm2135/sm2135.h:77]: `num_channels_` does not seem to be actually used anywhere?

File: tests/test3.yaml [hunk 4]
```diff
@@ -586,6 +586,10 @@ script:
     then:
       - lambda: 'ESP_LOGD("main", "Hello World!");'
 
+sm2135:
+  data_pin: GPIO12
+  clock_pin: GPIO14
+
```
Comment (jesserockz) [inline tests/test3.yaml:592]: Please also add some test lines for the `output`

**Review by jesserockz — CHANGES_REQUESTED** [review]
A few things below

File: esphome/components/sm2135/sm2135.cpp (hunk 2 above)
Comment (BoukeHaarsma23) [reply esphome/components/sm2135/sm2135.cpp:48]: Yes, the SM2135 always has 5 channels as I am aware of

File: esphome/components/sm2135/sm2135.h (hunk 3 above)
Comment (BoukeHaarsma23) [reply esphome/components/sm2135/sm2135.h:77]: Indeed, will remove!

File: tests/test3.yaml (hunk 4 above)
Comment (BoukeHaarsma23) [reply tests/test3.yaml:592]: Done!

**Review by jesserockz — APPROVED** [review]

File: esphome/components/sm2135/sm2135.cpp (hunk 2 above)
Comment (glmnet) [reply esphome/components/sm2135/sm2135.cpp:48]: Then this is no a configuration option, hence no point in printing it out?

Comment (BoukeHaarsma23) [reply esphome/components/sm2135/sm2135.cpp]: Yeah that makes sense. I have removed the line.

**Pull request merged**